Pytest4 errors - previously deprecated stuff is now Error

Related pages:
https://docs.pytest.org/en/latest/deprecations.html#cached-setup
https://github.com/pytest-dev/pytest/blob/master/CHANGELOG.rst#deprecations-1

AttributeError: 'SubRequest' object has no attribute 'cached_setup'
```
    @pytest.fixture
    def gw(request, execmodel):
        scope = request.config.option.scope
>       group = request.cached_setup(
            setup=execnet.Group,
            teardown=lambda group: group.terminate(timeout=1),
            extrakey="testgroup",
            scope=scope,
        )
E       AttributeError: 'SubRequest' object has no attribute 'cached_setup'
```

RemovedInPytest4Warning: getfuncargvalue is deprecated, use getfixturevalue
```
    @pytest.fixture
    def anypython(request):
        name = request.param
        executable = getexecutable(name)
        if executable is None:
            if sys.platform == "win32":
                executable = winpymap.get(name, None)
                if executable:
                    executable = py.path.local(executable)
                    if executable.check():
                        return executable
                    executable = None
            py.test.skip("no %s found" % (name,))
        if "execmodel" in request.fixturenames and name != 'sys.executable':
>           backend = request.getfuncargvalue("execmodel").backend
E           RemovedInPytest4Warning: getfuncargvalue is deprecated, use getfixturevalue

testing/conftest.py:152: RemovedInPytest4Warning
______________ ERROR at setup of test_popen_io[thread-python2.7] _______________
```

--- a/testing/conftest.py
+++ b/testing/conftest.py
@@ -133,7 +133,7 @@ def anypython(request):
                 executable = None
         py.test.skip("no {} found".format(name))
     if "execmodel" in request.fixturenames and name != 'sys.executable':
-        backend = request.getfuncargvalue("execmodel").backend
+        backend = request.getfixturevalue("execmodel").backend
         if backend != "thread":
             pytest.xfail(
                 "cannot run {!r} execmodel with bare {}".format(backend, name))
@@ -169,7 +169,7 @@ def gw(request, execmodel, group):
             gw.proxygw = proxygw
             assert pname in group
         elif request.param == "ssh":
-            sshhost = request.getfuncargvalue('specssh').ssh
+            sshhost = request.getfixturevalue('specssh').ssh
             # we don't use execmodel.backend here
             # but you can set it when specifying the ssh spec
             gw = group.makegateway("ssh={}//id=ssh".format(sshhost))

--- a/testing/conftest.py
+++ b/testing/conftest.py
@@ -40,23 +40,12 @@ def pytest_addoption(parser):
         '--gx', action="append", dest="gspecs", default=None,
         help="add a global test environment, XSpec-syntax. ")
     group.addoption(
-        '--gwscope', action="store", dest="scope", default="session",
-        type="choice", choices=["session", "function"],
-        help="set gateway setup scope, default: session.")
-    group.addoption(
         '--pypy', action="store_true", dest="pypy",
         help="run some tests also against pypy")
     group.addoption(
         '--broken-isp', action="store_true", dest="broken_isp",
         help=("Skips tests that assume your ISP doesn't put up a landing "
               "page on invalid addresses"))
-
-
-def pytest_report_header(config):
-    return [
-        "gateway test setup scope: %s" % config.getvalue("scope"),
-        "execnet: {} -- {}".format(execnet.__file__, execnet.__version__),
-    ]
 
 
 @pytest.fixture
@@ -151,15 +140,15 @@ def anypython(request):
     return executable
 
 
+@pytest.fixture(scope='session')
+def group():
+    g = execnet.Group()
+    yield g
+    g.terminate(timeout=1)
+
+
 @pytest.fixture
-def gw(request, execmodel):
-    scope = request.config.option.scope
-    group = request.cached_setup(
-        setup=execnet.Group,
-        teardown=lambda group: group.terminate(timeout=1),
-        extrakey="testgroup",
-        scope=scope,
-    )
+def gw(request, execmodel, group):
     try:
         return group[request.param]
     except KeyError:
@@ -188,6 +177,8 @@ def gw(request, execmodel):
             group.makegateway('popen//id=proxy-transport')
             gw = group.makegateway('popen//via=proxy-transport//id=proxy'
                                    '//execmodel=%s' % execmodel.backend)
+        else:
+            assert 0, "unknown execmodel: {}".format(request.param)
         return gw
 
 

--- a/testing/test_basics.py
+++ b/testing/test_basics.py
@@ -385,8 +385,10 @@ def test_remote_exec_function_with_kwargs(anypython, makegateway):
 
 def test_remote_exc__no_kwargs(makegateway):
     gw = makegateway()
-    pytest.raises(TypeError, gw.remote_exec, gateway_base, kwarg=1)
-    pytest.raises(TypeError, gw.remote_exec, 'pass', kwarg=1)
+    with pytest.raises(TypeError):
+        gw.remote_exec(gateway_base, kwarg=1)
+    with pytest.raises(TypeError):
+        gw.remote_exec('pass', kwarg=1)
 
 
 @skip_win_pypy

--- a/testing/test_channel.py
+++ b/testing/test_channel.py
@@ -37,7 +37,8 @@ class TestChannelBasicBehaviour:
 
     def test_channel_receive_timeout(self, gw):
         channel = gw.remote_exec('channel.send(channel.receive())')
-        pytest.raises(channel.TimeoutError, "channel.receive(timeout=0.2)")
+        with pytest.raises(channel.TimeoutError):
+            channel.receive(timeout=0.2)
         channel.send(1)
         channel.receive(timeout=TESTTIMEOUT)
 
@@ -263,9 +264,8 @@ class TestChannelBasicBehaviour:
         """)
         subchan = channel.receive()
         subchan.send(1)
-        excinfo = pytest.raises(
-            subchan.RemoteError,
-            "subchan.waitclose(TESTTIMEOUT)")
+        with pytest.raises(subchan.RemoteError) as excinfo:
+            subchan.waitclose(TESTTIMEOUT)
         assert "42" in excinfo.value.formatted
         channel.send(1)
         channel.waitclose()
@@ -289,7 +289,8 @@ class TestChannelFile:
         f = channel.makefile()
         assert not f.isatty()
         channel.waitclose(TESTTIMEOUT)
-        pytest.raises(IOError, f.write, 'hello')
+        with pytest.raises(IOError):
+            f.write('hello')
 
     def test_channel_file_proxyclose(self, gw):
         channel = gw.remote_exec("""

--- a/testing/test_gateway.py
+++ b/testing/test_gateway.py
@@ -138,9 +138,8 @@ class TestBasicGateway:
 
     def test_remote_exec_no_explicit_close(self, gw):
         channel = gw.remote_exec('channel.close()')
-        excinfo = py.test.raises(
-            channel.RemoteError,
-            "channel.waitclose(TESTTIMEOUT)")
+        with pytest.raises(channel.RemoteError) as excinfo:
+            channel.waitclose(TESTTIMEOUT)
         assert "explicit" in excinfo.value.formatted
 
     def test_remote_exec_channel_anonymous(self, gw):
@@ -218,9 +217,8 @@ class TestPopenGateway:
         assert x.lower() == str(waschangedir).lower()
 
     def test_remoteerror_readable_traceback(self, gw):
-        e = py.test.raises(
-            gateway_base.RemoteError,
-            'gw.remote_exec("x y").waitclose()')
+        with pytest.raises(gateway_base.RemoteError) as e:
+            gw.remote_exec("x y").waitclose()
         assert "gateway_base" in e.value.formatted
 
     def test_many_popen(self, makegateway):
@@ -253,9 +251,12 @@ class TestPopenGateway:
         """)
         remotepid = channel.receive()
         py.process.kill(remotepid)
-        py.test.raises(EOFError, "channel.waitclose(TESTTIMEOUT)")
-        py.test.raises(IOError, channel.send, None)
-        py.test.raises(EOFError, channel.receive)
+        with pytest.raises(EOFError):
+            channel.waitclose(TESTTIMEOUT)
+        with pytest.raises(IOError):
+            channel.send(None)
+        with pytest.raises(EOFError):
+            channel.receive()
 
     def test_receive_on_remote_sysexit(self, gw):
         channel = gw.remote_exec("""
@@ -371,8 +372,8 @@ class TestThreads:
 class TestTracing:
     def test_popen_filetracing(self, testdir, monkeypatch, makegateway):
         tmpdir = testdir.tmpdir
-        monkeypatch.setenv("TMP", tmpdir)
-        monkeypatch.setenv("TEMP", tmpdir)  # windows
+        monkeypatch.setenv("TMP", str(tmpdir))
+        monkeypatch.setenv("TEMP", str(tmpdir))  # windows
         monkeypatch.setenv('EXECNET_DEBUG', "1")
         gw = makegateway("popen")
         #  hack out the debuffilename

--- a/testing/test_multi.py
+++ b/testing/test_multi.py
@@ -176,14 +176,17 @@ class TestGroup:
         group = Group()
         gw = group.makegateway("popen//id=hello")
         assert group["hello"] == gw
-        py.test.raises((TypeError, AttributeError), "del group['hello']")
-        py.test.raises((TypeError, AttributeError), "group['hello'] = 5")
+        with pytest.raises((TypeError, AttributeError)):
+            del group['hello']
+        with pytest.raises((TypeError, AttributeError)):
+            group['hello'] = 5
         assert 'hello' in group
         assert gw in group
         assert len(group) == 1
         gw.exit()
         assert 'hello' not in group
-        py.test.raises(KeyError, "group['hello']")
+        with pytest.raises(KeyError):
+            _ = group['hello']
 
     def test_default_group(self):
         oldlist = list(execnet.default_group)

--- a/testing/test_rsync.py
+++ b/testing/test_rsync.py
@@ -90,12 +90,14 @@ class TestRSync:
         rsync.add_target(gw1, dirs.dest1)
         rsync.send()
         assert dirs.dest1.join('hello').check()
-        py.test.raises(IOError, "rsync.send()")
+        with pytest.raises(IOError):
+            rsync.send()
         assert rsync.send(raises=False) is None
         rsync.add_target(gw1, dirs.dest2)
         rsync.send()
         assert dirs.dest2.join('hello').check()
-        py.test.raises(IOError, "rsync.send()")
+        with pytest.raises(IOError):
+            rsync.send()
         assert rsync.send(raises=False) is None
 
     def test_rsync_default_reporting(self, capsys, dirs, gw1):

--- a/testing/test_serializer.py
+++ b/testing/test_serializer.py
@@ -116,12 +116,12 @@ def py3(request):
 
 @pytest.fixture(params=['py2', 'py3'])
 def dump(request):
-    return request.getfuncargvalue(request.param).dump
+    return request.getfixturevalue(request.param).dump
 
 
 @pytest.fixture(params=['py2', 'py3'])
 def load(request):
-    return request.getfuncargvalue(request.param).load
+    return request.getfixturevalue(request.param).load
 
 
 simple_tests = [

--- a/testing/test_xspec.py
+++ b/testing/test_xspec.py
@@ -23,7 +23,8 @@ class TestXSpec:
         assert spec.nice is None
         assert not hasattr(spec, '_xyz')
 
-        pytest.raises(AttributeError, "spec._hello")
+        with pytest.raises(AttributeError):
+            spec._hello()
 
         spec = XSpec("socket=192.168.102.2:8888//python=python2.5//nice=3")
         assert spec.socket == "192.168.102.2:8888"

